src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+            rgrid[1:ir_cut] .* proj .* drgrid[1:ir_cut]
+        end
+    end
+
+    PspUpf(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h, pswfcs,
+           pswfc_occs, rhoatom, rvlocpzdr, r2projsdr; identifier)
+end
+
+charge_ionic(psp::PspUpf) = psp.Zion
+
+"""
+    eval_psp_projector_real(psp::PspUpf, i::Number, l::Number, r::Number)
+
+Evaluate the ith Kleinman-Bylander β projector with angular momentum l at real-space
+distance r via linear interpolation on the real-space mesh.
+
+Note: UPFs store `r[i] p_{il}(r[i])`, so r=0 is undefined and will error.
+"""
+function eval_psp_projector_real(psp::PspUpf, i, l, r::T) where {T <: Real}
+    psp.projs_interp[l+1][i](r) / r
+end
+
+"""
+    eval_psp_projector_fourier(psp::PspUPF, i::Number, l::Number, q::Number)
+
+
+Evaluate the radial part of the i-th projector for angular momentum l at the
+reciprocal vector with modulus q:
+p(q) = ∫R^3 p{il}(r) e^{-iqr} dr
+     = 4π ∫{R+} r^2 p_{il}(r) j_l(q r) dr
+     = 4π Σ{i} r[i]^2 p_{il}(r[i]) j_l(q r[i]) dr[i]
+
+Note: UPFs store `r[i] p_{il}(r[i])`
+"""
+function eval_psp_projector_fourier(psp::PspUpf, i, l, q::T)::T where {T <: Real}
+    s = zero(T)
+    @inbounds for ir = eachindex(psp.r2projsdr[l+1][i])
+        s += sphericalbesselj_fast(l, q * psp.rgrid[ir]) * psp.r2projsdr[l+1][i][ir]
+    end
+    4T(π) * s
+end
+
+"""
+     eval_psp_local_real(psp::PspUpf, r::Number)
+
+Evaluate the local potential at real-space distance `r` via linear
+interpolation on the real-space mesh.
+ """
+eval_psp_local_real(psp::PspUpf, r::T) where {T <: Real} = psp.vloc_interp(r)
+
+"""
+    eval_psp_local_fourier(psp::PspUpf, q::Number)
+
+Evaluate the local part of the pseudopotential in reciprocal space
+using a Coulomb correction term -Z/r:
+V(q) = ∫{R^3} (Vloc(r) + Z/r) e^{-iqr} dr
+     = 4π ∫{R+} (Vloc(r) + Z/r) sin(qr)/qr r^2 dr
+     = 4π/q Σ{i} sin(q r[i]) (r[i] V(r[i]) + Z) dr[i]
+"""
+function eval_psp_local_fourier(psp::PspUpf, q::T)::T where {T <: Real}
+    s = zero(T)
+    @inbounds for ir = eachindex(psp.rvlocpzdr)
+        s += sin(q * psp.rgrid[ir]) * psp.rvlocpzdr[ir]
+    end
+    4T(π) * (s - psp.Zion / q) / q
+end
+
+"""
+    eval_psp_energy_correction(T::Type, psp::PspUpf, n_electrons::Number)
+
+Evaluate the energy correction to the Ewald electrostatic interaction energy of one unit
+cell, which is required compared the Ewald expression for point-like nuclei.
+n_electrons is the number of electrons per unit cell.
+This defines the uniform compensating background charge, which is assumed here.
+
+Notice: The returned result is the energy per unit cell and not the energy per volume.
+To obtain the latter, the caller needs to divide by the unit cell volume.
+
+The energy correction is defined as the limit of the Fourier-transform of the local
+potential as q -> 0:
+lim{q->0} 4π Nelec ∫{R+} (V(r) + Z/r) sin(qr)/qr r^2 dr
+= 4π Nelec ∫{R+} (V(r) + Z/r) r^2 dr
+= 4π Nelec Σ{i} r[i] (r[i] V(r[i]) + Z) dr[i]
```

Comment:
Don't repeat the comment from NCPSP, only comment on the specific UPF things. The formulas are usefu though, and should be moved to the NCPSP comment

Reply:
Also the comment in PspHgh.jl is relevant, ideally merge all these comments in the NCPSP one

Reply:
> should be moved to the NCPSP comment

or a NCPSP docstring, which we can parse and put on the dftk docs :)